### PR TITLE
Create a test help for writing simpler metadata tests

### PR DESCRIPTION
**Pull request by davidwengier**
Thoughts @tmat?

I like that it puts the source code near the expected results, but I wonder if you preferred it the other way?

Mainly it just stops me from doing bad copy-and-pastes where I copy from `diff1`, and missing changing one of the things to `reader2` or `md2` and get weird test failures 😛

### CODE REVIEW

**Comment by tmat** [comment]
Good idea. I'd however try to avoid having a single method (AddGeneration) with gazillion parameters that needs to express everything we may want to test. 

Instead, perhaps let it take a lambda that has access to the data and calls various Verify helpers on them.

**Comment by davidwengier** [comment]
Yeah, that could work, we'd just need AddGeneration to take the edits, and there could be a Verify method that takes a `Action<GenerationInfo>` that has the methods. Will have a play.

**Comment by davidwengier** [comment]
Updated with a lambda. I think I like it better, in particular its nice that I can now click the right thing on a stack trace in a test failure and be taken to the right verify call.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/MetadataTest.cs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+using System.Reflection.Metadata;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Emit;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.EditAndContinue.UnitTests
+{
+    internal partial class MetadataTest : IDisposable
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/MetadataTest.cs:19]: >MetadataTest [](http://example.com/codeflow?start=27&length=12)

Looks great, I like it! I'd name this type something more general through. We might want to validate IL and PDBs in additiona to metadata.

Perhpas `EditAndContinueTest` would do.

Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/MetadataTest.cs:19]: nit: sealed

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/MetadataTest.GenerationInfo.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        internal class GenerationInfo
+        {
+            private readonly string _description;
+            private readonly List<MetadataReader> _readers;
+
+            public readonly CSharpCompilation Compilation;
+            public readonly MetadataReader MetadataReader;
+            public readonly EmitBaseline Baseline;
+
+            public GenerationInfo(string description, List<MetadataReader> readers, CSharpCompilation compilation, MetadataReader reader, EmitBaseline baseline)
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/MetadataTest.GenerationInfo.cs:24]: >string description [](http://example.com/codeflow?start=34&length=18)

I'd just make this `int ordinal`.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/MetadataTest.GenerationInfo.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+using Microsoft.CodeAnalysis.Emit;
+
+namespace Microsoft.CodeAnalysis.CSharp.EditAndContinue.UnitTests
+{
+    internal partial class MetadataTest
+    {
+        internal class GenerationInfo
+        {
+            private readonly string _description;
+            private readonly List<MetadataReader> _readers;
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/MetadataTest.GenerationInfo.cs:18]: >List<MetadataReader> _readers; [](http://example.com/codeflow?start=29&length=30)

This is a bit odd - the class is immutable except for the list of readers.

Comment (davidwengier) [reply src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/MetadataTest.GenerationInfo.cs:18]: Yeah, I knew this was lazy when I did it, I should have fixed it. This should be a field that holds a reference to the test, and the readers should be publically accessible from there. Sometimes I get lazy with test code. Will fix :)

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/MetadataTest.cs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            _readers.Add(reader);
+
+            EncValidation.VerifyModuleMvid(_genId, _prevGeneration.MetadataReader, reader);
+
+            _prevGeneration = new GenerationInfo($"generation {_genId++}", _readers, compilation, md.Reader, diff.NextGeneration);
+
+            return this;
+        }
+
+        internal MetadataTest Verify(Action<GenerationInfo> verification)
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/MetadataTest.cs:80]: >Action<GenerationInfo> verification [](http://example.com/codeflow?start=37&length=35)

I'd pass the validator delegate to AddGeneration instead. Then make this `void Verify()` and run all validators.

That way you first build up the infos and then run validation.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/MetadataTest.GenerationInfo.cs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                var actual = _readers.GetStrings(MetadataReader.GetMemberRefNames());
+                AssertEx.Equal(expected, actual, message: $"MemberRefs don't match in {_description}");
+            }
+
+            internal void VerifyTableSize(TableIndex table, int expected)
+            {
+                AssertEx.AreEqual(expected, MetadataReader.GetTableRowCount(table), message: $"{table} table size doesnt't match in {_description}");
+            }
+
+            internal void VerifyEncLog(IEnumerable<EditAndContinueLogEntry> expected)
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/MetadataTest.GenerationInfo.cs:57]: >VerifyEncLog [](http://example.com/codeflow?start=26&length=12)

I have started moving tests to VerifyEncLogDefinitions and VerifyEncMapDefinitions, which exclude refs -- we are rarely interested in refs.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/MetadataTest.GenerationInfo.cs (hunk 3 above)
Comment (tmat) [reply src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/MetadataTest.GenerationInfo.cs:18]: So another thing - we should use the MetadataAggregator - see the helper that I added in my last PR.

Comment (davidwengier) [reply src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/MetadataTest.GenerationInfo.cs:18]: Will have to talk to you about this tomorrow. I tried using the `CheckNames` method you added, but it gets a different answer in my test for the type defs in the 3rd generation ¯\\\_(ツ)_/¯

**Comment by davidwengier** [comment]
Anything left on this @tmat?

Can always follow up with more validations as needed of course.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs [hunk 6]
```diff
@@ -362,207 +362,183 @@ public void ModifyMethod_WithTuples()
...
     static string F() { return null; }
-}";
-            var source1 =
-@"class C
+}",
+                    verification: g =>
+                    {
+                        g.VerifyTypeDefNames("<Module>", "C");
+                        g.VerifyMethodDefNames("Main", "F", ".ctor");
+                        g.VerifyMemberRefNames(/*CompilationRelaxationsAttribute.*/".ctor", /*RuntimeCompatibilityAttribute.*/".ctor", /*Object.*/".ctor", /*DebuggableAttribute*/".ctor", /*DescriptionAttribute*/".ctor");
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs:378]: >VerifyMemberRefNames [](http://example.com/codeflow?start=26&length=20)

As a follow up - we should be able to resolve fully qualified names of MemberRefs, MethodDefs and TypeDefs.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTestBase.cs [hunk 7]
```diff
@@ -193,23 +196,34 @@ public static void CheckNames(IList<MetadataReader> readers, ImmutableArray<Type
...
                 var genEntityHandle = aggregator.GetGenerationHandle(toHandle(handle), out int typeGeneration);
                 var nameHandle = getName(readers[typeGeneration], genEntityHandle);
 
-                var genNameHandle = (StringHandle)aggregator.GetGenerationHandle(nameHandle, out int nameGeneration);
-                return readers[nameGeneration].GetString(genNameHandle);
-            }));
+                return readers.GetString(nameHandle);
+
+                //var genNameHandle = (StringHandle)aggregator.GetGenerationHandle(nameHandle, out int nameGeneration);
+                //return readers[nameGeneration].GetString(genNameHandle);
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTestBase.cs:225]: This should work. Need to investigate if it doesn't

Comment (tmat) [reply src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTestBase.cs:225]: What test fails if you keep the original?

Comment (davidwengier) [reply src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTestBase.cs:225]: Oops, wasn't supposed to commit this. I was mucking around here trying to understand why the MetadataAggregator didn't work for the `VerifyTypeDefs` method. Will revert.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTest.cs [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+        private bool _hasVerified;
+
+        public EditAndContinueTest(CSharpCompilationOptions? options, TargetFramework? targetFramework)
+        {
+            _options = options;
+            _targetFramework = targetFramework ?? TargetFramework.Standard;
+        }
+
+        internal EditAndContinueTest AddGeneration(string source, Action<GenerationVerifier> verification)
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTest.cs:34]: Maybe we should name this parameter 'validator' across the board a la CompileAndVerify

https://github.com/dotnet/roslyn/blob/0c92c3f34bb6ea1e5f04d544670573a711beda52/src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs#L816-L828

Comment (davidwengier) [reply src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTest.cs:34]: Good idea.. I'm making this change right now in my browser. Hopefully committing it works!

**Review by tmat — APPROVED** [review]
:shipit:

**Pull request merged**